Reduce compilation times by forward-declaring DOMPromiseDeferred

jernoble opened the pull request:
#### 6211ce9e4d8de65866fd884f3b021f88b0aa5863
<pre>
Reduce compilation times by forward-declaring DOMPromiseDeferred
<a href="https://bugs.webkit.org/show_bug.cgi?id=251029">https://bugs.webkit.org/show_bug.cgi?id=251029</a>
rdar://104567742

Reviewed by Brent Fulgham.

Add a new forwarding header, JSDOMPromiseDeferredForward.h, that forward declares DeferredPromise,
DOMPromiseDeferred&lt;&gt;, and all the necssary IDL types which might be used in a header.

Use this new forwarding header everywhere, and move the JSDOMPromiseDeferred.h include into the
implementation files.

Elsewhere, classes declared ivars of these promise classes; those either require the class&apos;s constructors and
destructors to be moved to the implementation file, move std::optional -&gt; std::unique_ptr, or for the bare
promise to be wrapped in a UniqueRef.

These changes reduce the overall cost of including JSDOMPromiseDeferred to be reduced by 66%, from 300s of
compile time to 100s on this machine.

* Source/WebCore/Modules/WebGPU/GPU.cpp:
* Source/WebCore/Modules/WebGPU/GPU.h:
(WebCore::GPU::create):
* Source/WebCore/Modules/WebGPU/GPUAdapter.h:
* Source/WebCore/Modules/WebGPU/GPUBuffer.cpp:
* Source/WebCore/Modules/WebGPU/GPUBuffer.h:
* Source/WebCore/Modules/WebGPU/GPUDevice.cpp:
* Source/WebCore/Modules/WebGPU/GPUDevice.h:
* Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
* Source/WebCore/Modules/WebGPU/GPUQueue.h:
* Source/WebCore/Modules/WebGPU/GPUShaderModule.cpp:
* Source/WebCore/Modules/WebGPU/GPUShaderModule.h:
* Source/WebCore/Modules/applepay/ApplePaySetupWebCore.h:
* Source/WebCore/Modules/cache/DOMCacheStorage.cpp:
* Source/WebCore/Modules/credentialmanagement/BasicCredential.cpp:
* Source/WebCore/Modules/credentialmanagement/BasicCredential.h:
* Source/WebCore/Modules/encryptedmedia/MediaKeySystemRequest.h:
* Source/WebCore/Modules/fetch/FetchBody.cpp:
* Source/WebCore/Modules/fetch/FetchBody.h:
* Source/WebCore/Modules/fetch/FetchBodyConsumer.cpp:
(WebCore::FetchBodyConsumer::FetchBodyConsumer):
* Source/WebCore/Modules/fetch/FetchBodyConsumer.h:
(WebCore::FetchBodyConsumer::FetchBodyConsumer): Deleted.
* Source/WebCore/Modules/indexeddb/IDBFactory.cpp:
* Source/WebCore/Modules/indexeddb/IDBFactory.h:
* Source/WebCore/Modules/mediastream/PeerConnectionBackend.cpp:
* Source/WebCore/Modules/mediastream/RTCRtpSFrameTransform.cpp:
* Source/WebCore/Modules/mediastream/RTCRtpSFrameTransform.h:
* Source/WebCore/Modules/mediastream/RTCRtpScriptTransformer.h:
* Source/WebCore/Modules/mediastream/RTCRtpSender.h:
* Source/WebCore/Modules/push-api/PushManager.cpp:
* Source/WebCore/Modules/push-api/PushManager.h:
* Source/WebCore/Modules/push-api/PushSubscription.h:
* Source/WebCore/Modules/streams/ReadableStreamSource.cpp:
* Source/WebCore/Modules/streams/ReadableStreamSource.h:
* Source/WebCore/Modules/streams/WritableStreamSink.h:
* Source/WebCore/Modules/webaudio/AudioBuffer.h:
* Source/WebCore/Modules/webaudio/BaseAudioContext.cpp:
(WebCore::BaseAudioContext::decodeAudioData):
* Source/WebCore/Modules/webaudio/BaseAudioContext.h:
* Source/WebCore/Modules/webaudio/OfflineAudioContext.cpp:
* Source/WebCore/Modules/webaudio/OfflineAudioContext.h:
* Source/WebCore/Modules/webaudio/PeriodicWave.cpp:
* Source/WebCore/Modules/webaudio/WaveShaperNode.cpp:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.cpp:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.h:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.cpp:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.h:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.cpp:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.h:
* Source/WebCore/Modules/webxr/WebXRSession.cpp:
(WebCore::WebXRSession::didCompleteShutdown):
(WebCore::WebXRSession::end):
* Source/WebCore/Modules/webxr/WebXRSession.h:
* Source/WebCore/Modules/webxr/WebXRSystem.cpp:
* Source/WebCore/Modules/webxr/WebXRSystem.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/DOMCSSPaintWorklet.cpp:
* Source/WebCore/dom/AbortSignal.h:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
* Source/WebCore/page/WorkerNavigator.h:
* Source/WebCore/testing/ServiceWorkerInternals.cpp:
* Source/WebCore/testing/ServiceWorkerInternals.h:
* Source/WebCore/testing/WebFakeXRDevice.h:
* Source/WebCore/testing/WebXRTest.cpp:
* Source/WebCore/testing/WebXRTest.h:
* Source/WebCore/workers/service/FetchEvent.cpp:
* Source/WebCore/workers/service/FetchEvent.h:
* Source/WebCore/workers/service/ServiceWorkerContainer.h:
* Source/WebCore/workers/service/ServiceWorkerRegistration.h:
* Source/WebCore/worklets/Worklet.cpp:
* Source/WebCore/worklets/Worklet.h:

Canonical link: <a href="https://commits.webkit.org/259392@main">https://commits.webkit.org/259392@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5e0821815f817d77579d35aa0a43a7d32e9535f1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104736 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13813 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37639 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114011 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174210 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108653 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14944 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4745 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97066 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112934 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110496 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11560 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94567 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39097 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/108203 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93398 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26181 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80764 "Passed tests") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/94660 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7168 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27537 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/92627 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/4913 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7272 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4113 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30185 "Passed tests") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/103553 "Passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13322 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47097 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/101315 "Built successfully") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/6478 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9059 "Built successfully") | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/45/builds/25257 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/Modules/fetch/FetchBodyConsumer.h:
```diff
@@ -30,7 +30,7 @@
 
 #include "FetchBodySource.h"
 #include "FormDataConsumer.h"
-#include "JSDOMPromiseDeferred.h"
```
I wonder if we could `check-webkit-style` about `JSDOMPromiseDeferred.h` ever appearing in a header file?

jernoble replied:
We could. There are places where I left the full header in; mostly for entirely inlined header files and the like. But people doing that could ignore the style check.

brentfulgham commented on Source/WebCore/Headers.cmake:
```diff
@@ -526,6 +526,7 @@ set(WebCore_PRIVATE_FRAMEWORK_HEADERS
     bindings/js/JSDOMMicrotask.h
     bindings/js/JSDOMOperation.h
     bindings/js/JSDOMPromiseDeferred.h
+    bindings/js/JSDOMPromiseDeferredForward.h
```
Thank you for catching this!

brentfulgham commented on Source/WebCore/Modules/fetch/FetchBodyConsumer.cpp:
```diff
@@ -126,6 +126,15 @@ static std::optional<MimeType> parseMIMEType(const String& contentType)
+
+FetchBodyConsumer::FetchBodyConsumer(FetchBodyConsumer&&) = default;
+FetchBodyConsumer::~FetchBodyConsumer() = default;
+FetchBodyConsumer& FetchBodyConsumer::operator=(FetchBodyConsumer&&) = default;
```
Nit: I think some version of GTK will fail with default assignment operator, but I feel that the GTK port should add an #if/def for their old GCC version that lacks support (unlike their EWS bot, which *does* support it).

jernoble replied:
Agreed!

brentfulgham reviewed (approved):
Nice!

jernoble committed:
Reduce compilation times by forward-declaring DOMPromiseDeferred
https://bugs.webkit.org/show_bug.cgi?id=251029
rdar://104567742

Reviewed by Brent Fulgham.

Add a new forwarding header, JSDOMPromiseDeferredForward.h, that forward declares DeferredPromise,
DOMPromiseDeferred<>, and all the necssary IDL types which might be used in a header.

Use this new forwarding header everywhere, and move the JSDOMPromiseDeferred.h include into the
implementation files.

Elsewhere, classes declared ivars of these promise classes; those either require the class's constructors and
destructors to be moved to the implementation file, move std::optional -> std::unique_ptr, or for the bare
promise to be wrapped in a UniqueRef.

These changes reduce the overall cost of including JSDOMPromiseDeferred to be reduced by 66%, from 300s of
compile time to 100s on this machine.

* Source/WebCore/Modules/WebGPU/GPU.cpp:
* Source/WebCore/Modules/WebGPU/GPU.h:
(WebCore::GPU::create):
* Source/WebCore/Modules/WebGPU/GPUAdapter.h:
* Source/WebCore/Modules/WebGPU/GPUBuffer.cpp:
* Source/WebCore/Modules/WebGPU/GPUBuffer.h:
* Source/WebCore/Modules/WebGPU/GPUDevice.cpp:
* Source/WebCore/Modules/WebGPU/GPUDevice.h:
* Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
* Source/WebCore/Modules/WebGPU/GPUQueue.h:
* Source/WebCore/Modules/WebGPU/GPUShaderModule.cpp:
* Source/WebCore/Modules/WebGPU/GPUShaderModule.h:
* Source/WebCore/Modules/applepay/ApplePaySetupWebCore.h:
* Source/WebCore/Modules/cache/DOMCacheStorage.cpp:
* Source/WebCore/Modules/credentialmanagement/BasicCredential.cpp:
* Source/WebCore/Modules/credentialmanagement/BasicCredential.h:
* Source/WebCore/Modules/encryptedmedia/MediaKeySystemRequest.h:
* Source/WebCore/Modules/fetch/FetchBody.cpp:
* Source/WebCore/Modules/fetch/FetchBody.h:
* Source/WebCore/Modules/fetch/FetchBodyConsumer.cpp:
(WebCore::FetchBodyConsumer::FetchBodyConsumer):
* Source/WebCore/Modules/fetch/FetchBodyConsumer.h:
(WebCore::FetchBodyConsumer::FetchBodyConsumer): Deleted.
* Source/WebCore/Modules/indexeddb/IDBFactory.cpp:
* Source/WebCore/Modules/indexeddb/IDBFactory.h:
* Source/WebCore/Modules/mediastream/PeerConnectionBackend.cpp:
* Source/WebCore/Modules/mediastream/RTCRtpSFrameTransform.cpp:
* Source/WebCore/Modules/mediastream/RTCRtpSFrameTransform.h:
* Source/WebCore/Modules/mediastream/RTCRtpScriptTransformer.h:
* Source/WebCore/Modules/mediastream/RTCRtpSender.h:
* Source/WebCore/Modules/push-api/PushManager.cpp:
* Source/WebCore/Modules/push-api/PushManager.h:
* Source/WebCore/Modules/push-api/PushSubscription.h:
* Source/WebCore/Modules/streams/ReadableStreamSource.cpp:
* Source/WebCore/Modules/streams/ReadableStreamSource.h:
* Source/WebCore/Modules/streams/WritableStreamSink.h:
* Source/WebCore/Modules/webaudio/AudioBuffer.h:
* Source/WebCore/Modules/webaudio/BaseAudioContext.cpp:
(WebCore::BaseAudioContext::decodeAudioData):
* Source/WebCore/Modules/webaudio/BaseAudioContext.h:
* Source/WebCore/Modules/webaudio/OfflineAudioContext.cpp:
* Source/WebCore/Modules/webaudio/OfflineAudioContext.h:
* Source/WebCore/Modules/webaudio/PeriodicWave.cpp:
* Source/WebCore/Modules/webaudio/WaveShaperNode.cpp:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.cpp:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.h:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.cpp:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.h:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.cpp:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.h:
* Source/WebCore/Modules/webxr/WebXRSession.cpp:
(WebCore::WebXRSession::didCompleteShutdown):
(WebCore::WebXRSession::end):
* Source/WebCore/Modules/webxr/WebXRSession.h:
* Source/WebCore/Modules/webxr/WebXRSystem.cpp:
* Source/WebCore/Modules/webxr/WebXRSystem.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/DOMCSSPaintWorklet.cpp:
* Source/WebCore/dom/AbortSignal.h:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
* Source/WebCore/page/WorkerNavigator.h:
* Source/WebCore/testing/ServiceWorkerInternals.cpp:
* Source/WebCore/testing/ServiceWorkerInternals.h:
* Source/WebCore/testing/WebFakeXRDevice.h:
* Source/WebCore/testing/WebXRTest.cpp:
* Source/WebCore/testing/WebXRTest.h:
* Source/WebCore/workers/service/FetchEvent.cpp:
* Source/WebCore/workers/service/FetchEvent.h:
* Source/WebCore/workers/service/ServiceWorkerContainer.h:
* Source/WebCore/workers/service/ServiceWorkerRegistration.h:
* Source/WebCore/worklets/Worklet.cpp:
* Source/WebCore/worklets/Worklet.h:

Canonical link: https://commits.webkit.org/259392@main

darinadler commented on Source/WebCore/bindings/js/JSDOMPromiseDeferredForward.h:
```diff
@@ -0,0 +1,89 @@
+
+class DeferredPromise;
+
+template <typename IDLType> class DOMPromiseDeferred;
+template <typename IDLType> class DOMPromiseProxy;
```
I suggest omitting the space after `template` in a case like this.

darinadler commented on Source/WebCore/bindings/js/JSDOMPromiseDeferredForward.h:
```diff
@@ -0,0 +1,89 @@
+struct IDLUSVString;
+struct IDLObject;
+
+template<typename T> struct IDLInterface;
+template<typename T> struct IDLCallbackInterface;
+template<typename T> struct IDLCallbackFunction;
+template<typename T> struct IDLDictionary;
+template<typename T> struct IDLEnumeration;
+template<typename T> struct IDLNullable;
+template<typename T> struct IDLSequence;
+template<typename T> struct IDLFrozenArray;
+template<typename K, typename V> struct IDLRecord;
+template<typename T> struct IDLPromise;
```
These identifiers like `T`, `K`, and `V` can be omitted in a case like this.

darinadler commented on Source/WebCore/bindings/js/JSDOMPromiseDeferredForward.h:
```diff
@@ -0,0 +1,89 @@
+struct IDLError;
+struct IDLDOMException;
+
+template<typename... Ts> struct IDLUnion;
+template<typename T> struct IDLBufferSource;
```
`Ts` and `T` can be omitted.
